### PR TITLE
Set GCP to pull boot images from RHCOS pipeline.

### PR DESCRIPTION
**Pull request by patrickdillon**
Specifies a URL which points to a gcp storage bucket where the RHCOS pipeline publishes an RHCOS image for gcp (coreos/coreos-assembler#650). Use that tar.gz file to create a compute image for the cluster and boot the instances from that.

JIRA: [CORS-1164](https://jira.coreos.com/browse/CORS-1164)
/label platform/google

### CODE REVIEW

**Comment by abhinavdahiya** [comment]
This is missing the change on Machine objects... they should be using the image name https://github.com/openshift/installer/blob/470daaaf9952a72741eb2d807ca1f840f4466c4a/data/data/gcp/main.tf#L78

**Comment by csrwng** [comment]
Apart from updating machine objects, lgtm

**Comment by patrickdillon** [comment]
Updated machine spec here: https://github.com/openshift/installer/pull/2084/files#diff-8363d0f158df5ef13f50d06bad08fa0cR82

I think that handles it, let me know if I'm missing something

File: pkg/asset/machines/gcp/machines.go [hunk 1]
```diff
@@ -79,7 +79,7 @@ func provider(clusterID string, platform *gcp.Platform, mpool *gcp.MachinePool,
 			Boot:   true,
 			SizeGb: 128,
 			Type:   "pd-ssd",
-			Image:  osImage,
+			Image:  fmt.Sprintf("%s-rhcos-image", clusterID),
```
Comment (cgwalters) [inline pkg/asset/machines/gcp/machines.go]: Shouldn't we include e.g. the OpenShift major version here, like `4.1`, `4.2` etc?

Comment (abhinavdahiya) [reply pkg/asset/machines/gcp/machines.go]: As it stands today, we only have a single bootimage for all versions.. so this is appropriate.

Comment (cgwalters) [reply pkg/asset/machines/gcp/machines.go]: Namespacing with the cluster ID I think in general is OK.  But the reason I'm advocating including the version number is around issues stemming from https://github.com/openshift/os/issues/381

My concern is:

 - User sets up 4.2 cluster in GCP, cluster stays running
 - 4.3 is released
 - User later runs 4.3 installer which will have a new bootimage - if we end up overwriting that boot image, would e.g. newly scaled workers in the 4.2 cluster start using it? 

Basically I'm not sure we really want to think about supporting/dealing with 4.X clusters booting from 4.(X-1) etc.

Again I think the cluster ID here is OK, but think about e.g. bare metal PXE setups etc. where we don't programatially control the bootimages and may have multiple clusters at the same time; we want to encourage people in general to "version namespace" the bootimages.

Comment (abhinavdahiya) [reply pkg/asset/machines/gcp/machines.go]: > Namespacing with the cluster ID I think in general is OK. But the reason I'm advocating including the version number is around issues stemming from [openshift/os#381](https://github.com/openshift/os/issues/381)
> 
> My concern is:
> 
> * User sets up 4.2 cluster in GCP, cluster stays running
> * 4.3 is released
> * User later runs 4.3 installer which will have a new bootimage - if we end up overwriting that boot image, would e.g. newly scaled workers in the 4.2 cluster start using it?

The infra is is randomized for that purpose. So 2 clusters will have their own image.

> Basically I'm not sure we really want to think about supporting/dealing with 4.X clusters booting from 4.(X-1) etc.
> 
> Again I think the cluster ID here is OK, but think about e.g. bare metal PXE setups etc. where we don't programatially control the bootimages and may have multiple clusters at the same time; we want to encourage people in general to "version namespace" the bootimages.

I would suggest even those uses to create bootimage locations to be specific to the cluster.. they have access to infra-id

And as for versions in names, I don't think that are necessary here.

**Comment by patrickdillon** [comment]
I updated this to pull the image url from rhcos.json based on #2089. holding until that merges.

Currently I am including the commits from #2089, and will exclude them once it merges.

Based on https://github.com/openshift/installer/pull/2084#discussion_r307296483, I tried to pull in the image name (i.e. `rhcos-42-80-20190724-3` in this case) from rhcos.json and use that to create the image name, but it would be a slightly more significant rewrite than I want at that this stage.

File: data/data/gcp/main.tf [hunk 2]
```diff
@@ -16,7 +16,7 @@ module "bootstrap" {
 
   bootstrap_enabled = var.gcp_bootstrap_enabled
 
-  image_name   = var.gcp_image_id
+  image        = "${google_compute_image.cluster.self_link}"
```
Comment (abhinavdahiya) [inline data/data/gcp/main.tf]: this should follow terraform 0.12 syntax. `google_compute_image.cluster.self_link`

File: data/data/gcp/main.tf [hunk 3]
```diff
@@ -30,7 +30,7 @@ module "bootstrap" {
 module "master" {
   source = "./master"
 
-  image_name     = var.gcp_image_id
+  image          = "${google_compute_image.cluster.self_link}"
```
Comment (abhinavdahiya) [inline data/data/gcp/main.tf]: same as https://github.com/openshift/installer/pull/2084/files#r307397427

File: pkg/rhcos/gcp.go [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+	url := meta.GCP.URL
+	if url == "" {
+		return "", errors.New("no RHCOS GCP URL found")
+	}
+
+	return url, nil
+}
+
+// Image fetches the image name of the RHCOS image from the rhcos.json data
+func Image(ctx context.Context) (string, error) {
```
Comment (abhinavdahiya) [inline pkg/rhcos/gcp.go]: this is not used

File: pkg/rhcos/gcp.go [hunk 5]
```diff
@@ -0,0 +1,37 @@
+package rhcos
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+)
+
+// URL fetches the location of the public GCP storage bucket containing the RHCOS image
+func URL(ctx context.Context) (string, error) {
```
Comment (abhinavdahiya) [inline pkg/rhcos/gcp.go]: I think this should be called GCP 
`rhcos.GCP` becomes the API for getting GCP type image source.

File: pkg/asset/rhcos/image.go [hunk 6]
```diff
@@ -65,6 +65,7 @@ func (i *Image) Generate(p asset.Parents) error {
 		}
 		osimage, err = rhcos.AMI(ctx, config.Platform.AWS.Region)
 	case gcp.Name:
+		osimage = rhcos.GCP.URL
```
Comment (abhinavdahiya) [inline pkg/asset/rhcos/image.go]: this should use the `rhcos.GCP()`

**Comment by patrickdillon** [comment]
Removing hold now that #2094 has merged.
